docs/proposals/gpu-support.md

```diff
@@ -0,0 +1,339 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN MUNGE: GENERATED_TOC -->
+
+- [GPU support](#gpu-support)
+  - [Objective](#objective)
+  - [Background](#background)
+  - [Detailed discussion](#detailed-discussion)
+    - [Inventory](#inventory)
+    - [Scheduling](#scheduling)
+    - [The runtime](#the-runtime)
+      - [NVIDIA support](#nvidia-support)
+    - [Event flow](#event-flow)
+  - [Implementation plan](#implementation-plan)
+    - [V0](#v0)
+      - [Scheduling](#scheduling)
+      - [Runtime](#runtime)
+      - [Other](#other)
+  - [Future work](#future-work)
+    - [V1](#v1)
+    - [V2](#v2)
+    - [V3](#v3)
+    - [Undetermined](#undetermined)
+  - [Security considerations](#security-considerations)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+# GPU support
+
+Author: @therc
+
+Date: Apr 2016
+
+Status: Design in progress, early implementation of requirements
+
+## Objective
+
+Users should be able to request GPU resources for their workloads, as easily as
+for CPU or memory. Kubernetes should keep an inventory of machines with GPU
```

Comment:
@dchen1107 re: @davidopp 's  #19049 (comment) - totally agree. 

We used to call this arbitrary machine local limits, which other systems copied the basic design:  ( xref https://htcondor-wiki.cs.wisc.edu/index.cgi/tktview?tn=2905 ) 

Step 1: on startup there would be an option to kick a "plugin|discovery_script" whose output would be name value pairs of capabilities to be used for scheduling.    

Step 2: On allocate/match there would be another accompanying plugin/script to provision/assign said resource to a job, in this case would be pod and the machines status would be refreshed to ensure we decremented the resource count so the scheduler did the appropriate thing. 

In being generic you can apply arbitrary resource labeling.  (gpus/fpgas/infiniband cards/booster rockets/ion cannons/etc.) ;-)